event-protocol/README.md

```diff
@@ -0,0 +1,150 @@
+# Cucumber Event Protocol
+
+The Cucumber event protocol is a data exchange format for components in the
+Cucumber ecosystem.
+
+When Cucumber runs, it produces (emits) Cucumber events that are consumed by
+formatters that report the results in various formats. The protocol is designed
+to allow for other producers to emit events, and for other consumers to consume
+them.
+
+The protocol is transport agnostic, and can be transmitted in-process as well as
+between processes. This means some components (such as the [HTML Formatter](../html-formatter/README.md))
+can be implemented in a single programming language and used by Cucumber
+implementations in different languages.
+
+The Cucumber event protocol can also be used to implement parallel and distributed
+execution of Cucumber scenarios.
+
+Programs that support the protocol can be producers or consumers (or both).
+
+Examples of producers are:
+
+* Cucumber
+* [Gherkin-Lint](../gherkin-lint/README.md)
+
+Examples of consumers are:
+* Formatters (HTML, JUnit, Progress, PDF)
+* 3rd party tools (Cucumber Pro)
+
+## Events {#events}
+
+### start {#event-start}
+
+Example:
+[import](examples/events/001_start.json)
+
+### source {#event-source}
+
+Example:
+[import](examples/events/002_source.json)
+
+### attachment {#event-attachment}
+
+An `attachment` attaches a piece of metadata to a particular line in a `source` file.
+Attachments can have many types (specified by a media type), and are typically used for:
+
+* `text/vnd.cucumber.step.status+plain`, status of a failing step (passed, undefined, pending, failed, skipped)
+* `text/vnd.cucumber.stacktrace.java+plain`, stack trace of a failing step (from java in this case)
+* `image/png`, `image/jpg`, `image/gif` - screenshots
+
+Example:
+[import](examples/events/003_attachment-png-embedded.json)
+
+## Implementation
+
+* Cucumber events are formatted as [Newline Delimited JSON](http://ndjson.org)
+* The transport protocol can use one or more of the following:
+  * `STDIN`/`STDOUT`
+  * Raw socket
+  * W3C WebSocket
+  * W3C EventSource (each event is a JSON-encoded event without `event:` type - it's in the payload itself)
+
+### Philosophy
+
+As the event specification evolves to support a richer set of events there are some
+key principles to consider:
+
+#### File format agnostic
+
+While Gherkin is currently the only file format that will be used in `source`
+events, no events should assume that all files will be Gherkin documents. This
+allows for alternative document formats in the future.
+
+#### No Cucumber execution semantics
+
+Some consumers (such as the [HTML Formatter](../html-formatter/README.md)) may
+consume events emitted by other consumers than Cucumber (for example [Gherkin-Lint](../gherkin-lint/README.md)).
+
+For this reason, events containing information about execution should be represented
+in a more generic way.
```

Comment:
When you open a connection to this TCP socket and start to produce the event stream then you can identify all events coming from this connection with a unique series, username and more. And then they are multiplexed they can have this unique identifier with them. Like:

``` json
{
  "user": "john",
  "series": "df1d3970-644e-11e6-8b77-86f30ca893d3",
  "event": {
    type: "start",
    timestamp: 1471420027068
  }
}
```

Events can be wrapped into more context. Because their environment is different. I think composition here is a better answer. Running a console formatter is always an unique stream of events and there is no point in having the series. But when you start to multiplex these streams to a long running daemon, only then it becomes necessary to have more context, if it is a cloud where users could track their test run history or a local server, these are different and have more context compared to the console formatter.